Design/ISIS_Imaging/ISIS_Imaging_general.md

```diff
@@ -0,0 +1,181 @@
+
+## Loading
+Loading of the images must be done via the `core.imgdata` module, which already handles loading of any supported image types. Any future extensions for new file types must be a part of the core module.
+
+The Loading will be done via a dialogue:
+- It will ask for the paths for Sample, Flat and Dark images. Only Sample path is required and loaded. Dark and Flat images are NOT loaded, but the paths will be cached.
+- The user can specify indices in 3 spinboxes for [Start, Stop, Step]. The spinboxes cannot be negative. Boundaries are: 
+    - Max value of Stop is the number of files with the extension (detected dynamically after selection of the file). 
+        - Stop does not have a lower bound, meaning the user can select a Stop value lower than the Start value. Doing so will result in a ValueError. 
+        - This can be avoided with a signal triggered by changing the Start spinbox, that changes the setMinimum of the Stop spinbox. (low priority)
+    - Max value of Start is Stop - 1. 
+    - Min value of Step is 1, and max value of Step is the value of Stop.
+- The selection will require the user to select a file, NOT a directory.
+- The image format will be determined from the selected file.
+- The absolute path to the directory will be determined from the selected file.
+
+## Handling multiple stacks in dialogues
+This is a consequence of being able to load multiple stack inside the program. How does the user select on which stack to apply the operation? 
+
+This produces problems when trying to sync in after a window has been closed. The naive approach of keeping them in a list does not work, as it would require every stack window to know it's position inside the list, but if the one that is closed is _not_ the last, they all get out of sync. Thus it's going to be a nightmare to implement and test.
+
+The current desing for the solution is to generate a unique ID for each stack, and store that plus a reference to the stack object in a dictionary. Qt provides `closeEvent` which is triggered when the user clicks the `X` button. Each stack window will know it's own unique ID, and then deletion (and sync) is just a `del dict[uid]`. 
+
+To answer the initial question, building the selection for the user is now just a matter of iterating the dict and getting the (user friendly) name, not the uuid, and displaying that.
+
+The unique ID will be generated using python's `uuid` package.
+
+A future upgrade on this will take into account which is the _current focus_ of the user, and when applying an operation will automatically select _that_ stack as the default choice.
+
+## Visualisation
+Visualisation will be done using Matplotlib. Each stack will be a separate part of a QDockWidget, so the users will be able to arrange the visualised stacks in any way they want. This also means that there can be more than one stack loaded at a time.
+
```

Comment:
To come back to this question, Matplotlib with PyQt5 seems to do the visualisation about 5x faster for 2048x2048 images. It has been mentioned a possibility for 4096x4096 images in the future. 

I've scaled some images up to that size, and the native drawing algorithms seem to be performing well. It is not quite as fast as `ImageJ/FIJI`, but maybe there is a potential for optimisation.

I also scaled some images up to 8192x8192 and 16386x16386. Both matplotlib and ImageJ seem to struggle with that size, which I think is expected. We can still load them, but changing to another image takes ~1 second for 8k and ~5 seconds for 16k